haskell.compiler.ghc*: fix cross-built native GHC

AlexandreTunstall opened the pull request:
###### Description of changes

This PR fixes all versions of GHC older than 9.6 to allow cross-compiling GHC itself (build ≠  host = target). In addition to that, it adds an overridable option to force building an unregisterised version.

Changes have also been made to 9.6 and newer, but they are merely tentative because Hadrian cannot currently cross-compile GHC.

Due to bugs in cc-wrapper and binutils-wrapper, cross-compiling GHC still doesn't produce binaries usable without emulation or further fixes, but this will at the very least allow users to port GHC to platforms that Nixpkgs doesn't support natively.

###### Things done


- Built on platform(s)
  - [X] x86_64-linux
    - [X] `pkgsCross.aarch64-multiplatform.haskell.compiler.integer-simple.ghc{884,8107}`
    - [X] `pkgsCross.aarch64-multiplatform.haskell.compiler.native-bignum.ghc{902,924,925,926,927,928,942,943,944,945}`
    - [X] `pkgsCross.riscv64.haskell.compiler.integer-simple.ghc884`
    - [X] `pkgsCross.riscv64.haskell.compiler.native-bignum.ghc945`
    - [X] `haskell.compiler.native-bignum.ghc945`
    - [X] `haskell.compiler.native-bignum.ghc962`
    - There are way too many affected derivations for me to try all of them; it's time-consuming and gobbles up tons of disk space.
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [N/A] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
  - [X] x86_64-linux
    - [X] `pkgsCross.aarch64-multiplatform.haskell.compiler.integer-simple.ghc{884,8107}`
    - [ ] `pkgsCross.aarch64-multiplatform.haskell.compiler.native-bignum.ghc{902,924,925,926,927,928,942,943,944}`
    - [X] `pkgsCross.aarch64-multiplatform.haskell.compiler.native-bignum.ghc945`
    - [X] `pkgsCross.riscv64.haskell.compiler.integer-simple.ghc884`
    - [X] `pkgsCross.riscv64.haskell.compiler.native-bignum.ghc945`
    - [X] `haskell.compiler.native-bignum.ghc945`
    - [X] `haskell.compiler.native-bignum.ghc962`
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

amjoseph-nixpkgs commented:
Hi, thanks for writing this PR!

This is a really really big commit (not your fault; ghc still uses [copy-paste-tweak style instead of switch-on-version style](https://github.com/NixOS/nixpkgs/pull/243607)) which makes it very hard to review; would you mind breaking it up into a separate commit for each mechanical change (like "search and replace FOO with BAR" or "replace libffi with targetPackages.libffi")?  Then we can focus our review effort on the remaining commits with non-mechanical changes, which should be short and focused.

Also, have you tried building anything that uses Template Haskell?  Last time I tried to get `build,host=x86_64-linux` `target=aarch64-linux` to work using `qemu-user` for Template Haskell, I hit a known GHC bug in their custom linker that caused segfaults.  I can dig up the GHC bug link if you like.

Unfortunately Template Haskell is widely used in the Haskell ecosystem; there are a couple of pervasive dependencies (like Vty) that use it, so everything that uses those needs it.

sternenseemann commented:
> Also, have you tried building anything that uses Template Haskell? Last time I tried to get `build,host=x86_64-linux` `target=aarch64-linux` to work using `qemu-user` for Template Haskell, I hit a known GHC bug in their custom linker that caused segfaults. I can dig up the GHC bug link if you like.

That is kind of orthogonal to this PR which gets `build != host, host == target` to work properly (I recall that building with build, host, target distinct is impossible). `-fexternal-interpreter` is something we should still tackle, but it makes more sense to do so in a different context, I'd say.

amjoseph-nixpkgs commented:
> That is kind of orthogonal to this PR which gets `build != host, host == target` to work properly

Ah, so, first line of the PR should probably be adjusted to:

> This PR fixes all versions of GHC older than 9.6 to allow ~both building GHC as a cross-compiler and~ cross-compiling GHC itself.

AlexandreTunstall commented:
> would you mind breaking it up into a separate commit for each mechanical change (like "search and replace FOO with BAR" or "replace libffi with targetPackages.libffi")?

@amjoseph-nixpkgs I assume by this you mean split it up by change (e.g. apply X change to each version) and not split it up by version (e.g. fix 9.4.5), right? Sure.

> Also, have you tried building anything that uses Template Haskell?

I have encountered a few errors that mentioned `-fexternal-interpreter` while trying to use a riscv64 cross-compiled GHC to bootstrap GHC 9.6.2 natively, but so far they've all been in test suites (so I disabled checks to get around them). My goal with this isn't to build a fully-functional GHC, but just something to bootstrap GHC on otherwise unsupported platforms. As far as I'm aware, building GHC does not require TH support.

I haven't tried using the cross-compilers to build any TH-using code. As sternenseemann mentioned, this is more about fixing cross-compiling GHC itself. I assume Nixpkgs already supports building GHC as a cross-compiler, but if it doesn't, this PR also makes that possible.

amjoseph-nixpkgs commented:
> @amjoseph-nixpkgs I assume by this you mean split it up by change (e.g. apply X change to each version) and not split it up by version (e.g. fix 9.4.5), right? Sure.

Correct!


> My goal with this isn't to build a fully-functional GHC, but just something to bootstrap GHC on otherwise unsupported platforms.

Ah, thanks, now I get it!  Yes this will be very useful, because the ancient GHC-supplied binaries we've been using to bootstrap on `powerpc64le` have ceased working.  So I'd like to use this there too.  Awesome!

AlexandreTunstall commented:
`pkgsCross.aarch64-multiplatform.pkgsBuildHost.haskell.compiler.native-bignum.ghc945` seems to build fine on Nixpkgs already, so I've amended the PR description.

AlexandreTunstall committed:
haskell.compiler.ghc8107: support {LD,AR}_STAGE0

AlexandreTunstall committed:
haskell.compiler.ghc*: set stage 0 tools

AlexandreTunstall committed:
haskell.compiler.ghc*: fix libffi cross dependency

AlexandreTunstall committed:
haskell.compiler.ghc*: fix ncurses splicing

AlexandreTunstall committed:
haskell.compiler.ghc*: don't configure --host

AlexandreTunstall committed:
haskell.compiler.ghc*: use target CC in settings

AlexandreTunstall committed:
haskell.compiler.ghc*: don't cross-compile haddock

AlexandreTunstall committed:
haskell.compiler.ghc*: add enableUnregisterised

AlexandreTunstall committed:
haskell.compiler.ghc*: assert cross is possible

AlexandreTunstall committed:
haskell.compiler.ghc*: allow all platforms

Propagating the supported platforms of the boot compiler doesn't make
much sense when unregistered cross-compilation is possible.

AlexandreTunstall committed:
haskell.compiler.ghc*: use GHC from pkgsBuildBuild

AlexandreTunstall commented:
@amjoseph-nixpkgs I have split this into 11 separate (hopefully atomic) commits.

I've also spent much of my week trying to compile GHC 9.6.2 natively on RISC-V using an unregisterised GHC 9.4.5 cross-compiled using these changes, and I've figured out what it takes.

```nix
pkgs.haskell.compiler.ghc962.override (old: rec {
  bootPkgs = pkgs.haskell.packages.ghc945.override {
    buildHaskellPackages = bootPkgs;

    ghc = let
      passthru = {
        targetPrefix = "";
        enableShared = false;
        hasHaddock = false;

        llvmPackages = pkgs.llvmPackages_12;

        haskellCompilerName = "ghc-9.4.5";
      };
    in passthru // {
      version = "9.4.5";

      outPath = builtins.storePath boot/ghc;

      inherit passthru;

      meta = {
        license = lib.licenses.bsd3;
        platforms = [ "riscv64-linux" ];
      };
    };

    overrides = self: super: {
      mkDerivation = args: super.mkDerivation ({
        enableLibraryProfiling = false;
      } // args);

      # These test suites don't compile with boot GHC
      alex = pkgs.haskell.lib.compose.dontCheck super.alex;
      data-array-byte = pkgs.haskell.lib.compose.dontCheck super.data-array-byte;
      doctest = pkgs.haskell.lib.compose.dontCheck super.doctest;
      hashable = pkgs.haskell.lib.compose.dontCheck super.hashable;
      optparse-applicative = pkgs.haskell.lib.compose.dontCheck super.optparse-applicative;
      QuickCheck = pkgs.haskell.lib.compose.dontCheck super.QuickCheck;
      temporary = pkgs.haskell.lib.compose.dontCheck super.temporary;
      vector = pkgs.haskell.lib.compose.dontCheck super.vector;
    };
  };
})
```

And I needed to patch Nixpkgs so that Hadrian wouldn't try to use RTS flags only supported by the threaded runtime.
I couldn't work out how to do this with `override`/`overrideAttrs`.

```patch
diff --git a/pkgs/development/tools/haskell/hadrian/default.nix b/pkgs/development/tools/haskell/hadrian/default.nix
index 5911c34982b..da4d194e220 100644
--- a/pkgs/development/tools/haskell/hadrian/default.nix
+++ b/pkgs/development/tools/haskell/hadrian/default.nix
@@ -29,6 +29,8 @@ mkDerivation {
     # Additionally we need to recompile it on every change of UserSettings.hs.
     # See https://gitlab.haskell.org/ghc/ghc/-/merge_requests/1190
     "-O0"
+    # Don't use threaded-only RTS flags at runtime
+    "-f-threaded"
   ];
   isLibrary = false;
   isExecutable = true;
```

I'd like to make it possible to use a non-threaded GHC to build GHC, but I'm not sure how. Should it be as above? Should it be an overridable option to GHC which is propagated to Hadrian? Should it be a flag on GHC derivations similar to `hasHaddock`? Anyway, that's a change for another PR.

AlexandreTunstall commented:
Is it possible to have Hydra cross-compile GHC for all platforms?

It could potentially simplify bootstrapping on unsupported platforms by substituting the derivation instead of having to manually build it on a supported platform, copy the closure to the host, and write some rather elaborate code to convince Nixpkgs to use it as a boot compiler.

In fact, we could even configure Nixpkgs to do so by default on unsupported platforms, so that users can just `nix-shell -p ghc` and get a working compiler.

sternenseemann commented:
> It could potentially simplify bootstrapping on unsupported platforms by substituting the derivation instead of having to manually build it on a supported platform, copy the closure to the host, and write some rather elaborate code to convince Nixpkgs to use it as a boot compiler.

We can via `release-cross.nix`. Relying on substitution would be bad form here, rather we should maintain a derivation that downloads a known good build artefact using an ordinary fetcher, fixes runtime dependencies with `patchelf` etc. Not sure yet if we need to upload something to `tarballs.nixos.org` for this or we can fetch from the normal binary cache for this.

sternenseemann reviewed (changes requested):
Sorry I took so long to get back to you! This looks great, I'll be sure to play around with it a bit as well. It is also somehow simpler than I expected, e.g. it is not necessary to set any make variables telling it to build stage 2?

All suggestions are only noted once, since the changes are duplicated (very neatly, I must say) over the expressions.

I normally dislike tentative changes that don't work yet, but I guess the changes to the hadrian expression are alright as the retain some measure of consistency with the old expressions!

Also since this PR was opened, `9.4.6.nix` was added, so we'll need to remember to take care of that at the end.

sternenseemann commented:
> We can via `release-cross.nix`.

To clarify, one option would be to use `release-cross.nix` which [tests](https://hydra.nixos.org/jobset/nixpkgs/cross-trunk) the `master` branch and add it here:

https://github.com/NixOS/nixpkgs/blob/a14013769370b021e23200e7199d8cfaeb97098a/pkgs/top-level/release-cross.nix#L45-L50

The question is of course if this makes sense for all platforms `gnuCommon` is applied to.

Alternatively we can make a new section in `release-haskell.nix` which [tests](https://hydra.nixos.org/jobset/nixpkgs/haskell-updates) the `haskell-updates` branch. We tests some obscure GHC variants there already and it is monitored more closely by the Haskell maintainers, so probably a better fit. Also changes usually go through haskell-updates first…

sternenseemann commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -400,7 +432,8 @@ stdenv.mkDerivation (rec {
     ] ++ lib.teams.haskell.members;
     timeout = 24 * 3600;
-    inherit (ghc.meta) license platforms;
+    platforms = lib.platforms.all;
```
This change is right, but means we need to find a new mechanism to implement this: https://github.com/NixOS/nixpkgs/pull/240348. Not sure how that should look tbh, there is not a lot of prior art on detecting if something can be boostrapped in nixpkgs…

AlexandreTunstall replied:
Ideally, I think we'd want a way of recursively checking that the entire closure supports the build platform and isn't marked broken. I'm not aware of an existing function to check that.

Perhaps it'd be easier to give users the option to override `writeShellApplication` to disable the usage of ShellCheck.

If we still want to automatically choose whether to use it or not, how about this?

1. Check if ShellCheck is marked broken.
2. Check if the GHC used for ShellCheck is marked broken or unavailable.
3. Check if the boot GHC used is marked broken or unavailable (repeat until we get to a pre-compiled GHC).

It's not general enough to be usable on anything in Nixpkgs, but it should at the very least prevent a regression here.

sternenseemann commented:
> I'd like to make it possible to use a non-threaded GHC to build GHC, but I'm not sure how. Should it be as above?

An extra flag in the arguments to the hadrian expression should be fine.

For the `dontCheck`s we can use the `bootstrapPackageSet` overlay in `haskell-packages.nix` would be best if those overrides are only applied if they need to be.

But agreed, let's do that in a separate PR!

AlexandreTunstall commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -250,6 +251,9 @@ stdenv.mkDerivation (rec {
 
   postPatch = "patchShebangs .";
 
+  # GHC is unable to build a cross-compiler without this set.
```
Unfortunately I have no idea myself.

I figured out it was needed by diffing the environment variables between `stdenv.mkDerivation` and `pkgsBuildTarget.stdenv.mkDerivation` (IIRC).

AlexandreTunstall commented on pkgs/development/compilers/ghc/9.2.4.nix:
```diff
@@ -61,6 +61,13 @@
+  # option will force it to do an unregistered build when set to true.
+  # See https://gitlab.haskell.org/ghc/ghc/-/wikis/building/unregisterised
+  # Registerised RV64 cross-compiler currently produces programs that segfault
+  enableUnregisterised ? !stdenv.buildPlatform.isRiscV64 && stdenv.targetPlatform.isRiscV64
```
In my testing, I found that a registerised stage 1 cross-compiler outputs programs that segfault, meaning that both stage 1 (outputs broken programs) and stage 2 (is itself broken) are unusable. I don't see the need to check `hostPlatform`, since `targetPlatform` is `RiscV64` whether we want stage 1 or 2.

There isn't an upstream issue yet, but I've been meaning to create one. I'm currently doing a native build of 9.2.8 to verify that it isn't already fixed in 9.6.2 (the only version I've tested on native so far). I'll add a link to it in the code once I create it.

AlexandreTunstall replied:
I've since discovered that native RISC-V compilers also segfault, so I'll change this to `stdenv.hostPlatform.isRiscV64 || stdenv.targetPlatform.isRiscV64` (`hostPlatform` to ensure stage 2 GHC doesn't crash, `targetPlatform` to ensure compiled Haskell programs don't crash).

AlexandreTunstall replied:
And I'll also include a link to a GHC issue.

amjoseph-nixpkgs commented:
Sorry it took me so long to get to this.

> Due to bugs in cc-wrapper and binutils-wrapper

Details?  Bug report?

amjoseph-nixpkgs reviewed (commented):
> @amjoseph-nixpkgs I have split this into 11 separate (hopefully atomic) commits.

Awesome!  Thanks, this is nice and clean.

Sorry it took me so long to get to this.  LGTM except:

1. I have not yet built this (will do that tonight).
2. The first commit, which runs `autoreconfHook` for some versions and patches `./configure` for others.  This creates a real footgun.  Let's run `autoreconfHook` everywhere and not patch `./configure`.
3. > Due to bugs in cc-wrapper and binutils-wrapper

   Details?  Bug report?  A long-form explanation of what has been accomplished and what does/doesn't work should go in the commit message for the last commit.

4. > cross-compiling GHC still doesn't produce binaries usable without emulation or further fixes, but this will at the very least allow users to port GHC to platforms that Nixpkgs doesn't support natively.

   I'm having trouble understanding this.  If the cross-built GHC doesn't work, how does it "allow users to port GHC to platforms that Nixpkgs doesn't support natively"?

5. > Is it possible to have Hydra cross-compile GHC for all platforms?

   Yes, but we should probably be respectful of the computing load.

   Please add an entry to `tests.cross.sanity` for a good `build=x86_64` example of a build that should succeed with this PR?  That will guard against anybody undoing all your hard work later on!

amjoseph-nixpkgs commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -226,6 +226,12 @@ stdenv.mkDerivation (rec {
+    (fetchpatch {
+      url = "https://gitlab.haskell.org/ghc/ghc/-/commit/8f7dd5710b80906ea7a3e15b7bb56a883a49fed8.patch";
+      hash = "sha256-C636Nq2U8YOG/av7XQmG3L1rU0bmC9/7m7Hty5pm5+s=";
+    })
```
> because that patch patches `configure`, not `configure.ac`, and `8.10.7` uses `autoreconfHook` (but `8.8.4` doesn't).

Note that the copy-paste-tweak style that our GHC expression uses makes these kinds of footguns inevitable.  You have to stare at the code for a long time to realize when somebody did something six times but the seventh time they did it differently.

@sternenseemann you guys really should consider doing something like this to the GHC expression:

- https://github.com/NixOS/nixpkgs/pull/249707

...otherwise it's way too easy for bugs to slip in during all-versions refactorings like this PR.

AlexandreTunstall replied:
I'd have used `autoreconfHook` everywhere, but it unfortunately didn't work for 8.8.4. Patching `configure` was the simplest fix I could come up with.

AlexandreTunstall replied:
By the way, the configure patch is redundant now that 8.8 has been removed, so I have removed it altogether.

sternenseemann commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -278,6 +282,9 @@ stdenv.mkDerivation (rec {
     export CLANG="${buildTargetLlvmPackages.clang}/bin/${buildTargetLlvmPackages.clang.targetPrefix}clang"
   '' + ''
+    export CC_STAGE0="${buildCC}/bin/${buildCC.targetPrefix}cc"
+    export LD_STAGE0="${buildCC.bintools}/bin/${buildCC.bintools.targetPrefix}ld${lib.optionalString useLdGold ".gold"}"
```
You can now (#239247), but we want to use gold even if it is available, but not the default; so the setting wouldn't be global.

AlexandreTunstall commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -326,7 +326,8 @@ stdenv.mkDerivation (rec {
     "--datadir=$doc/share/doc/ghc"
-    "--with-curses-includes=${ncurses.dev}/include" "--with-curses-libraries=${ncurses.out}/lib"
+    "--with-curses-includes=${pkgsBuildHost.ncurses.dev}/include"
+    "--with-curses-libraries=${pkgsBuildHost.ncurses.out}/lib"
```
It is supposedly disabled, but something still ends up using ncurses.

AlexandreTunstall commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -1,8 +1,8 @@
 # build-tools
 , bootPkgs
-, autoconf, automake, coreutils, fetchpatch, fetchurl, perl, python3, m4, sphinx
+, autoreconfHook, autoconf, automake, coreutils, fetchpatch, fetchurl, perl, python3, m4, sphinx
```
The latest `haskell-updates` has dropped 8.8.4, so I don't think there will be much reason to do this once I've rebased. After rebasing, all the commit will do is add `autoreconfHook` and a patch for `configure.ac`.

AlexandreTunstall commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -48,24 +48,33 @@
-assert (stdenv.targetPlatform != stdenv.hostPlatform) -> !enableHaddockProgram;
+assert (stdenv.buildPlatform != stdenv.hostPlatform || stdenv.targetPlatform != stdenv.hostPlatform) -> !enableHaddockProgram;
+
+# GHC does not support building when all 3 platforms are different.
```
Noted, though I agree about the future-proofing (for both the comment and the asserts).

AlexandreTunstall commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -400,7 +432,8 @@ stdenv.mkDerivation (rec {
     ] ++ lib.teams.haskell.members;
     timeout = 24 * 3600;
-    inherit (ghc.meta) license platforms;
+    platforms = lib.platforms.all;
```
Is there documentation about this? `git grep -wne 'hydraPlatform' -- '*.md'` doesn't find anything and looking for "platform" gives far too many matches.

AlexandreTunstall commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -302,18 +322,20 @@ stdenv.mkDerivation (rec {
       done
   '';
 
+  # We don't pass --host because ./configure gets confused when we do
```
This is quite a tricky issue to explain, so I'll elaborate the comment with a slightly better explanation followed by a link to the relevant GHC wiki page.

AlexandreTunstall committed:
tests.cross.sanity: Add GHC to catch regressions

This is to ensure that Haskell users on platforms that lack official
bindists still have a convenient means of getting GHC running natively.

In my admittedly somewhat limited testing on RISC-V, GHC 8.10.7 is able
to bootstrap native builds for 9.2.8 and 9.4.5. GHC 9.2.8 and 9.4.5 are
unable to bootstrap themselves and 9.6.2 when cross-compiled.

If you're looking at this commit to see whether you can safely upgrade
the compiler used here to remove 8.10, please try cross-compiling 9.0 or
later and then booting a native GHC with it.

AlexandreTunstall commented:
> The first commit, which runs autoreconfHook for some versions and patches ./configure for others. This creates a real footgun. Let's run autoreconfHook everywhere and not patch ./configure.

This is no longer necessary due to the removal of 8.8.4.

> Details? Bug report? A long-form explanation of what has been accomplished and what does/doesn't work should go in the commit message for the last commit.

See #173952, which is a fix for binutils-wrapper. cc-wrapper has the same problem: they reference the build platform's shell in the wrapper script when they should be referencing the host platform's shell.

This results in GHC not being runnable on the host platform without emulating the build platform.

> I'm having trouble understanding this. If the cross-built GHC doesn't work, how does it "allow users to port GHC to platforms that Nixpkgs doesn't support natively"?

Emulation can be used as a workaround.

> Please add an entry to tests.cross.sanity for a good build=x86_64 example of a build that should succeed with this PR? That will guard against anybody undoing all your hard work later on!

Done. I haven't limited it to x86_64, as I assume that cross-compiling should work on any build platform.

skeuchel commented:
This is great. Thanks for doing this! I hope we can use this to get "trusted" bootstrap tarballs for new platforms (riscv64) without waiting for GHCHQ to release bindists.

I have not yet built any of it but looked over some of the changes.

#### Unregisterised riscv64 builds
  This is a pity but of course unavoidable in general for new platforms. I would still like to get newer e.g. >= 9.6 registerised builds working if possible. I think @AlexandreTunstall had some trouble with registerised cross-compiled ghcs. Can you comment what exactly went wrong and which versions you tried? In my experience **native** registerised builds for >= 9.6 work out of the box. Also Debian ships with a **native** 9.4.7 registerised build which I believe they got working by [building against llvm 15](https://salsa.debian.org/haskell-team/DHG_packages/-/blob/7e045c97427d5be1f1174c855a887540613df46c/p/ghc/debian/patches/newer-llvm). Might be useful if you do any testing since that is less painful than starting from an unregisterised one. Here is the info of that build:
```
$ ghc --info | egrep -i 'cross|boot|link|llvm|interp|ffi|register|ways'
 ,("C compiler link flags","-fuse-ld=bfd")
 ,("cross compiling","NO")
 ,("target has RTS linker","NO")
 ,("Unregisterised","NO")
 ,("LLVM target","riscv64-unknown-linux")
 ,("LLVM llc command","llc-15")
 ,("LLVM opt command","opt-15")
 ,("LLVM clang command","clang")
 ,("Use interpreter","YES")
 ,("RTS ways","debug thr thr_debug thr_p dyn debug_dyn thr_dyn thr_debug_dyn thr_debug_p debug_p")
 ,("Use LibFFI","YES")
 ,("Booter version","9.4.7")
 ,("Have interpreter","YES")
 ,("Target default backend","LLVM")
```

#### Interpreter in hadrian builds
The move to hadrian initially disabled the interpreter. This is a bit minor if you just want a compiler to bootstrap. There are two upstream MR enabling that again:
- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10714 (merged)
- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12286 (pending)

skeuchel commented on pkgs/development/compilers/ghc/common-hadrian.nix:
```diff
@@ -164,18 +165,33 @@
+  # See https://gitlab.haskell.org/ghc/ghc/-/wikis/building/unregisterised
+  # Registerised RV64 compiler produces programs that segfault
+  # See https://gitlab.haskell.org/ghc/ghc/-/issues/23957
+  enableUnregisterised ? stdenv.hostPlatform.isRiscV64 || stdenv.targetPlatform.isRiscV64
```
Registerised riscv64 builds of ghc >=9.6 work for me natively (build = host = target) and this is the default behaviour before this PR. Can you preserve that for all hadrian-based builds?

AlexandreTunstall replied:
Is that true even when building more complex programs like Pandoc?

From my testing with 9.6.2, registerised initially appeared to work until I tried compiling something more complex.

AlexandreTunstall replied:
I'll give building with LLVM 15 a try.

In the meantime, I'll undo this change in `common-hadrian.nix`.

AlexandreTunstall replied:
For older versions of GHC, LLVM 15 is not officially supported, so I don't want to take the risk of fixing registerised builds.

Especially now that 9.6 is the default version.

skeuchel replied:
> Is that true even when building more complex programs like Pandoc?
> 
> From my testing with 9.6.2, registerised initially appeared to work until I tried compiling something more complex.

Yes I got pandoc and shellcheck installed and tested basic functionality. Only cachix failed to compile with "error: cycle detected in build of '/nix/store/[..]-cachix-1.7.drv' in the references of output 'bin' from output 'out'" which I have yet to look into. I will try to rebuild a new registerised 9.6.4 from your branch and will report back.

AlexandreTunstall replied:
9.6.2 still segfaults on my old (~August) NixOS build when using LLVM 15.

```
[ 1 of 16] Compiling Language.Haskell.HsColour.Classify ( Language/Haskell/HsColour/Classify.hs, dist/build/Language/Haskell/HsColour/Classify.o, dist/bu>
/nix/store/mjlk72gj834v6lmbk9j9pvby88s686cz-stdenv-linux/setup: line 1597:   202 Segmentation fault      (core dumped) ./Setup build
```

This was bootstrapped using a native unregisterised 9.2.8, which was itself booted from a cross-compiled 8.10.7 compiled with the original version of this PR. All native packages were compiled for `RV64GC_Zba_Zbb` using `gcc13Stdenv`.

Some (likely useless) kernel logs:

```
ghc_worker[772846]: unhandled signal 11 code 0x1 at 0x0000004e3584b582
CPU: 2 PID: 772846 Comm: ghc_worker Not tainted 6.5.0 #1-NixOS
Hardware name: StarFive VisionFive 2 v1.3B (DT)
epc : 0000004e3584b582 ra : 0000003ff3e06900 sp : 00000004e3ffa730
 gp : 000000000006ea80 tp : 00000004e3fff8e0 t0 : 0000000000004000
 t1 : 0000003ff3e00628 t2 : 0000003ff6f86598 s0 : 00000004f148d652
 s1 : 0000003fedb8afd8 a0 : 0000000000000000 a1 : 0000003fed8dd038
 a2 : 0000000000000002 a3 : 0000000000000002 a4 : 0000000000000032
 a5 : 09e1854e3584b583 a6 : 0000000000000000 a7 : 0000003ff7ffdd50
 s2 : 00000004f311b9f8 s3 : 00000004f16d93c8 s4 : 0000003fed8dc5e0
 s5 : 00000004f31140c0 s6 : 00000004f148d638 s7 : 00000004f36d2870
 s8 : 00000004f2ce1000 s9 : 00000000000003bd s10: 0000003fedb59058
 s11: 00000004f31140c0 t3 : 0000003fed2f0514 t4 : 0000000000000001
 t5 : 0000003fed9c6e2a t6 : 000000000000002e
status: 8000000200006020 badaddr: 0000004e3584b582 cause: 000000000000000c
```

I'll try 9.6.4 on a more up-to-date Nixpkgs to see if the issue still happens.

AlexandreTunstall replied:
I discovered that the 6.5 kernel has a regression that causes some applications (notably rustc) to segfault. I've upgraded my kernel to 6.6, which doesn't have that issue, but the GHC 9.6.2 I've previously built still segfaults.

(The 9.6.4 build is still ongoing.)

AlexandreTunstall replied:
The 9.6.4 build is fully working. I have successfully built ShellCheck and nix-output-monitor with it, and not a single segfault has been seen.

Now I need to try rebuilding 9.6.2 to properly rule out the Linux kernel as the cause and try building 9.2 and 9.4 registerised to see if I can get those to work. This is going to take a lot of time.

skeuchel commented:
So I cross-compiled `integer-simple.ghc8107` and `native-bignum.ghc{902,925,926,927,928,945,946,947,948}` from x86_64 to riscv64. Everything build just fine. However, there are x86_64 specific store paths leaking into the result. I tried to test basic functionality on a riscv64 machine.

The host platform's llvm is referenced. Is that even used for an unregisterised build?
```
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.10.7
$ ghc --info | grep -i llvm
 ,("LLVM target","riscv64-unknown-linux")
 ,("LLVM llc command","/nix/store/vcxvkk05f3fb3175bk8b9r4pa849l6cd-llvm-12.0.1/bin/llc")
 ,("LLVM opt command","/nix/store/vcxvkk05f3fb3175bk8b9r4pa849l6cd-llvm-12.0.1/bin/opt")
 ,("LLVM clang command","clang")
$ file /nix/store/vcxvkk05f3fb3175bk8b9r4pa849l6cd-llvm-12.0.1/bin/llc
/nix/store/vcxvkk05f3fb3175bk8b9r4pa849l6cd-llvm-12.0.1/bin/llc: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/1rm6sr6ixxzipv5358x0cmaw8rs84g2j-glibc-2.38-44/lib/ld-linux-x86-64.so.2, BuildID[sha1]=8a085589d1caa5f6d7eb2bc59ec420fe188151d7, for GNU/Linux 3.10.0, not stripped
$ file /nix/store/vcxvkk05f3fb3175bk8b9r4pa849l6cd-llvm-12.0.1/bin/opt
/nix/store/vcxvkk05f3fb3175bk8b9r4pa849l6cd-llvm-12.0.1/bin/opt: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/1rm6sr6ixxzipv5358x0cmaw8rs84g2j-glibc-2.38-44/lib/ld-linux-x86-64.so.2, BuildID[sha1]=7903cd1677264d41f42a61aa0faec1b20827042e, for GNU/Linux 3.10.0, not stripped
```

Compiling a simple hello world fails because it references the host platform's C compiler
```
$ ghc --info | grep 'C compiler command'
 ,("C compiler command","/nix/store/haai65bgzwfq3m954f3148yl76y2wgcc-gcc-wrapper-13.2.0/bin/cc")
$ ghc --make main.hs
[1 of 1] Compiling Main             ( main.hs, main.o )
/nix/store/haai65bgzwfq3m954f3148yl76y2wgcc-gcc-wrapper-13.2.0/bin/cc: runInteractiveProcess: posix_spawnp: invalid argument (Exec format error)
$ head -n1 /nix/store/haai65bgzwfq3m954f3148yl76y2wgcc-gcc-wrapper-13.2.0/bin/cc
#! /nix/store/5lr5n3qa4day8l1ivbwlcby2nknczqkq-bash-5.2p26/bin/bash
$ file /nix/store/5lr5n3qa4day8l1ivbwlcby2nknczqkq-bash-5.2p26/bin/bash
/nix/store/5lr5n3qa4day8l1ivbwlcby2nknczqkq-bash-5.2p26/bin/bash: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /nix/store/1rm6sr6ixxzipv5358x0cmaw8rs84g2j-glibc-2.38-44/lib/ld-linux-x86-64.so.2, BuildID[sha1]=3b212210bbdd58aef6576efc0c8bf8808e0ff53d, for GNU/Linux 3.10.0, not stripped
```

AlexandreTunstall commented:
> However, there are x86_64 specific store paths leaking into the result.

Ouch, that is correct. Fortunately, it is possible to work around those issues by ensuring that you have a suitable LLVM in your `PATH`.

```
[nixos@nixos:~/system]$ /etc/nixos/boot/ghc-8.10.7/bin/ghc --info
 ,("LLVM llc command","/nix/store/ivvw5czh7yl0flc3q0hi45v7hqfqirww-llvm-12.0.1/bin/llc")
 ,("LLVM opt command","/nix/store/ivvw5czh7yl0flc3q0hi45v7hqfqirww-llvm-12.0.1/bin/opt")

[nixos@nixos:~/system]$ /nix/store/ivvw5czh7yl0flc3q0hi45v7hqfqirww-llvm-12.0.1/bin/llc --version
-bash: /nix/store/ivvw5czh7yl0flc3q0hi45v7hqfqirww-llvm-12.0.1/bin/llc: cannot execute binary file: Exec format error

[nixos@nixos:~/system]$ ghc --info | grep -e llvm
 ,("LLVM llc command","/nix/store/3x5fkjh20mcf4lhs0ji297x8avlj81py-llvm-12.0.1/bin/llc")
 ,("LLVM opt command","/nix/store/3x5fkjh20mcf4lhs0ji297x8avlj81py-llvm-12.0.1/bin/opt")

[nixos@nixos:~/system]$ export PATH=/nix/store/3x5fkjh20mcf4lhs0ji297x8avlj81py-llvm-12.0.1/bin:$PATH

[nixos@nixos:~]$ cd ~/ghc-test/

[nixos@nixos:~/ghc-test]$ /etc/nixos/boot/ghc-8.10.7/bin/ghc Hello.hs
[1 of 1] Compiling Main             ( Hello.hs, Hello.o )
Linking Hello ...

[nixos@nixos:~/ghc-test]$ ./Hello
Hello, world!
```

I can't remember why I didn't fix that when I first wrote the PR, so I'll look into it.

AlexandreTunstall commented:
Actually, that might be a regression introduced while rebasing (the GHC I have here was built before rebasing). All the more reason to investigate.

```
[nixos@nixos:~/ghc-test]$ /etc/nixos/boot/ghc-8.10.7/bin/ghc Hello.hs
[1 of 1] Compiling Main             ( Hello.hs, Hello.o )
Linking Hello ...

[nixos@nixos:~/ghc-test]$ llc --version
-bash: llc: command not found

[nixos@nixos:~/ghc-test]$ gcc --version
-bash: gcc: command not found
```

AlexandreTunstall commented:
> Compiling a simple hello world fails because it references the host platform's C compiler

This is because of the aforementioned issue with `cc-wrapper` and `binutils-wrapper`. The C compiler used is the host platform's, but the wrappers use the build platform's bash, causing it to fail.

I have a (potentially hacky) fix in my `ghc-cross-usable` branch. I'm not looking to merge it in this PR as it's a completely unrelated change.

AlexandreTunstall commented:
> Can you comment what exactly went wrong and which versions you tried?

I have no fresh memories of trying older versions, but there are more details in the corresponding GHC issue: https://gitlab.haskell.org/ghc/ghc/-/issues/23957

> The [build] platform's llvm is referenced. Is that even used for an unregisterised build?

You're probably right about this, which is why I never noticed that the LLVM in the `ghc --info` isn't correct.

I tried my earlier example of a working cross-compiled GHC without LLVM in PATH and it still worked. That GHC was built with patched cc-wrapper and binutils-wrapper.

skeuchel commented:
To test the changes to 9.4.8 and the tentative changes made to 9.6, I've successfully tested the following

- Building `#{git-annex,shellcheck,pandoc}` (so using 9.6.4) and `#haskell.packages.ghc948.{git-annex,ShellCheck,pandoc}` on x86_64 from this PR's commit https://github.com/NixOS/nixpkgs/commit/4e0921f5217f18d92373b4c9d5a8b9ee0ad62085.
- Overridden the ghc-9.6.3 binary of https://github.com/NixOS/nixpkgs/commit/6f791b432bd07e33f27a738f0d2f171f3a28de15 to use a registerised 9.6.3 bindist which I self-compiled on debian; overridden the 9.6.4 derivation to use 9.6.3 as the boot compiler; added the upstream commits https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10714 and https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12286 as patches to 9.6.4; and built git-annex, shellcheck and pandoc using 9.6.4.

This is what I currently care about ;)

skeuchel commented:
As for the wrappers, looking through the code I can see that the settings file
https://github.com/NixOS/nixpkgs/blob/4e0921f5217f18d92373b4c9d5a8b9ee0ad62085/pkgs/development/compilers/ghc/9.4.8.nix#L378-L389 is supposed to be populated with the host platform's tools (`installCC` comes from `pkgsHostTarget`). 

Do I understand correctly that the wrappers used in `pkgsHostTarget` are to blame? So (without looking at the details) this is fixed by https://github.com/NixOS/nixpkgs/pull/173952 for binutils?

If this is the case, then this seems a bit orthogonal and I would not block the PR because of this, but I'll let the maintainer decide.

Also is the following still true or outdated?
https://github.com/NixOS/nixpkgs/blob/08b9151ed40350725eb40b1fe96b0b86304a654b/pkgs/development/compilers/ghc/9.4.8.nix#L168-L170

sternenseemann commented:
https://github.com/NixOS/nixpkgs/pull/305392 has been merged into #339272. Further testing feedback/testing appreciated. Currently ironing out some regressions.
